src/UglyToad.PdfPig/DocumentLayoutAnalysis/NearestNeighbourWordExtractor .cs

```diff
@@ -0,0 +1,199 @@
+            });
+
+            // Group indexes
+            List<List<int>> groupedIndexes = new List<List<int>>();
+            List<int> indexDone = new List<int>();
+            for (int c = 0; c < lettersCount; c++)
+            {
+                int i = indexes[c];
+                if (i == -1) continue;
+
+                bool isDoneC = indexDone.Contains(c);
+                bool isDoneI = indexDone.Contains(i);
+                if (isDoneC || isDoneI)
+                {
+                    if (isDoneC && !isDoneI)
+                    {
+                        foreach (var pair in groupedIndexes.Where(x => x.Contains(c)))
+                        {
+                            pair.Add(i);
+                        }
+                        indexDone.Add(i);
+                    }
+                    else if (!isDoneC && isDoneI)
+                    {
+                        foreach (var pair in groupedIndexes.Where(x => x.Contains(i)))
+                        {
+                            pair.Add(c);
+                        }
+                        indexDone.Add(c);
+                    }
+                    else
+                    {
+                        foreach (var pair in groupedIndexes.Where(x => x.Contains(i)))
+                        {
+                            if (!pair.Contains(c)) pair.Add(c);
+                        }
+
+                        foreach (var pair in groupedIndexes.Where(x => x.Contains(c)))
+                        {
+                            if (!pair.Contains(i)) pair.Add(i);
+                        }
+                    }
+                }
+                else
+                {
+                    List<int> pair = new List<int>() { c, i };
+                    groupedIndexes.Add(pair);
+                    indexDone.AddRange(pair);
+                }
+            }
+
+            // Merge lists with common index 
+            for (int c = 0; c < lettersCount; c++)
+            {
+                List<List<int>> candidates = groupedIndexes.Where(x => x.Any(t => t == c)).ToList();
+                if (candidates.Count < 2) continue; // only one group with this index
+
+                List<int> merged = candidates.First();
+                groupedIndexes.Remove(merged);
+                for (int i = 1; i < candidates.Count; i++)
+                {
+                    var current = candidates[i];
+                    merged = merged.Union(current).ToList();
+                    groupedIndexes.Remove(current);
+                }
+                groupedIndexes.Add(merged);
+            }
+
+            List<Word> words = new List<Word>();
+            for (int a = 0; a < groupedIndexes.Count(); a++)
+            {
+                List<Letter> groupedLetters = new List<Letter>();
+                foreach (int s in groupedIndexes[a])
+                {
+                    groupedLetters.Add(letters[s]);
+                }
+                words.Add(new Word(groupedLetters));
```

Comment:
Is it possible for letters to be out of order at this point? I think Linq operations don't guarantee order so even though I think in normal flow the letters will be in order it might be worth an explicit orderby here?

Reply:
Not sure if possible but will check that